src/compiler/checker.ts

```diff
@@ -5351,17 +5355,40 @@ module ts {
                 }
                 // Target type is type of prototype property
                 let prototypeProperty = getPropertyOfType(rightType, "prototype");
-                if (!prototypeProperty) {
-                    return type;
+                if (prototypeProperty) {
+                    let targetType = getTypeOfSymbol(prototypeProperty);
+                    if (targetType !== anyType) {
+                        // Narrow to the target type if it's a subtype of the current type
+                        if (isTypeSubtypeOf(targetType, type)) {
+                            return targetType;
+                        }
+                        // If the current type is a union type, remove all constituents that aren't subtypes of the target.
+                        if (type.flags & TypeFlags.Union) {
+                            return getUnionTypeOfSubtypeConstituents(<UnionType>type, targetType);

```

Comment:
I'm not a huge fan of this name. I've talked it over with @mhegazy and @DanielRosenwasser, and it's probably better to keep this line as it was, instead of factoring it out.